packages/core/test/financial-templates/perpetual-multiparty/PositionManager.js

```diff
@@ -0,0 +1,1459 @@
+    // pay > 100% fees on the PfC. This should pay a maximum of 100% of the PfC without reverting.
+    const pfc = await positionManager.pfc();
+    const feesOwed = (
+      await store.computeRegularFee(startTime.addn(1), startTime.addn(102), { rawValue: pfc.toString() })
+    ).regularFee;
+    assert(Number(pfc.toString()) < Number(feesOwed.toString()));
+    const farIntoTheFutureSeconds = 502;
+    await positionManager.setCurrentTime(startTime.addn(farIntoTheFutureSeconds));
+    const payTooManyFeesResult = await positionManager.payRegularFees();
+    truffleAssert.eventEmitted(payTooManyFeesResult, "RegularFeesPaid", ev => {
+      // There should be 98.99 + 0.99 = 99.98 collateral remaining in the contract.
+      return ev.regularFee.toString() === toWei("99.98") && ev.lateFee.toString() === "0";
+    });
+    assert.equal((await positionManager.getCollateral(sponsor)).toString(), "0");
+
+    // TODO: Add unit tests for when the latePenalty > 0 but (latePenalty + regularFee > pfc). The component fees need to be reduced properly.
+
+    // Set the store fees back to 0 to prevent it from affecting other tests.
+    await store.setFixedOracleFeePerSecondPerPfc({ rawValue: "0" });
+
+    // Check that no event is fired if the fees owed are 0.
+    await positionManager.setCurrentTime(startTime.addn(farIntoTheFutureSeconds + 1));
+    const payZeroFeesResult = await payRegularFees();
+    truffleAssert.eventNotEmitted(payZeroFeesResult, "RegularFeesPaid");
+  });
+
+  it("Emergency shutdown: lifecycle", async function() {
+    // Create one position with 100 synthetic tokens to mint with 150 tokens of collateral. For this test say the
+    // collateral is Dai with a value of 1USD and the synthetic is some fictional stock or commodity.
+    await collateral.approve(positionManager.address, toWei("100000"), { from: sponsor });
+    const numTokens = toWei("100");
+    const amountCollateral = toWei("150");
+    await positionManager.create({ rawValue: amountCollateral }, { rawValue: numTokens }, { from: sponsor });
+
+    // Transfer half the tokens from the sponsor to a tokenHolder. IRL this happens through the sponsor selling tokens.
+    const tokenHolderTokens = toWei("50");
+    await tokenCurrency.transfer(tokenHolder, tokenHolderTokens, { from: sponsor });
+
+    // Some time passes and the UMA token holders decide that Emergency shutdown needs to occur.
+    const shutdownTimestamp = Number(await positionManager.getCurrentTime()) + 1000;
+    await positionManager.setCurrentTime(shutdownTimestamp);
+
+    // Should revert if emergency shutdown initialized by non-FinancialContractsAdmin (governor).
+    assert(await didContractThrow(positionManager.emergencyShutdown({ from: other })));
+
+    // FinancialContractAdmin can initiate emergency shutdown.
+    await financialContractsAdmin.callEmergencyShutdown(positionManager.address);
```

Comment:
Thoughts on setting a final fee and testing that the final fee is paid by the gov?